timemachine/md/smc.py

```diff
@@ -0,0 +1,179 @@
+    log_prob: BatchLogProb,
+    resample: Resampler,
+) -> ResultDict:
+    """barebones implementation of Sequential Monte Carlo (SMC)
+
+    Parameters
+    ----------
+    samples: [N,] list
+    lambdas: [K,] array
+    propagate: function
+        [move(x, lam) for x in xs]
+        for example, move(x, lam) might mean "run 100 steps of all-atom MD targeting exp(-u(., lam)), initialized at x"
+    log_prob: function
+        [exp(-u(x, lam)) for x in xs]
+    resample: function
+        (optionally) perform resampling given an array of log weights
+
+    Returns
+    -------
+    trajs_dict
+        "sample_traj"
+            [K-1, N] list of snapshots
+        "incremental_log_weights_traj"
+            [K-1, N] array of incremental log weights
+        "ancestry_traj"
+            [K-1, N] array of ancestor idxs
+        "log_weights_traj"
+            [K, N] array of accumulated log weights
+
+    References
+    ----------
+    * Arnaud Doucet's annotated bibliography of SMC
+        https://www.stats.ox.ac.uk/~doucet/smc_resources.html
+    * [Dai, Heng, Jacob, Whiteley, 2020] An invitation to sequential Monte Carlo samplers
+        https://arxiv.org/abs/2007.11936
+    """
+    n = len(samples)
+    log_weights = np.zeros(n)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+        sample_traj.append(samples)
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+
+    # main loop
+    for (lam_initial, lam_target) in zip(lambdas[:-2], lambdas[1:-1]):
+        # update log weights
+        incremental_log_weights = log_prob(sample_traj[-1], lam_target) - log_prob(sample_traj[-1], lam_initial)
+        log_weights += incremental_log_weights
+
+        # resample
+        indices, log_weights = resample(log_weights)
+        resampled = [sample_traj[-1][i] for i in indices]
+
+        # propagate
+        samples = propagate(resampled, lam_target)
+
+        # log
+        accumulate_results(samples, indices, log_weights, incremental_log_weights)
+
+    # final result: a collection of samples, with associated log weights
+    incremental_log_weights = log_prob(samples, lambdas[-1]) - log_prob(samples, lambdas[-2])
+    incremental_log_weights_traj.append(incremental_log_weights)
+    log_weights_traj.append(np.array(log_weights + incremental_log_weights))
+
+    # cast everything (except samples list) to arrays
+    trajs_dict = dict(
+        traj=sample_traj,
+        log_weights_traj=np.array(log_weights_traj),
+        ancestry_traj=np.array(ancestry_traj),
+        incremental_log_weights_traj=np.array(incremental_log_weights_traj),
+    )
+    return trajs_dict
```

Comment:
nit: should we prefer returning results as `NamedTuple`s vs `dict`s in general? (this would allow for more static checking)

Reply:
I preferred to use a dict (with simple key and value types) over a namedtuple here for pickle convenience, esp. while the code is rapidly changing.

Can pickle with this version of timemachine, and if I change my mind about the field names or add fields or whatever in a later PR, I may still be able to load pickle files written with older versions (or different branches) of timemachine. However, if I make similar changes to the namedtuple definition between versions, reading / writing pickles between versions would be less convenient.

Reply:
Clarifications from offline discussion:
* changing the _class name_ of the namedtuple between versions versions used for pickle.dump and pickle.load would cause pickle.load to fail
* changing the _number_ of fields in the namedtuple between versions would cause pickle.load to fail
* if only the _names_ of fields changed, pickle.load would not have any errors
* if only the _order_ of fields changed, pickle.load would not have any errors, but would invoke the namedtuple constructor with arguments in the wrong order

`Dict[str, array]`, while less amenable to verification, seems more convenient for allowing pickle to "just work" between timemachine versions that are likely to add/remove/reorder fields.

In offline discussion, @mcwitt suggested possibly using pydantic or another tool for schema versioning, which I think would be a safer and better long-term solution for situations like this.